api/agent.go

```diff
@@ -0,0 +1,59 @@
+package api
+
+import (
+	"encoding/json"
+	"syscall"
+
+	ocispecs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+type AgentConfig struct {
+	// ConnProto: the method to connect to agent(or proxy)
+	// "unix": the paths are unix socket
+	// "vsock": use cid and port pair to connect to the agent (kata-only)
+	// "kvmtool": the paths are kvmtool driver created pty slaver with escaping character
+	ConnProto string
+	// AgentProto: the agent protocol
+	// "hyperstart": orginal hyperstart protocol
+	// "kata": kata agent protocol
+	// "kata-yamux": kata agent protocol over yamux
+	AgentProto string
+
+	HyperstartCtlPath, HyperstartStreamPath string
+	KataPath                                string
+	KataVsockCid, KataVsockPort             uint32
+}
+
+type Agent interface {
+	Config() AgentConfig
+	Realese() (json.RawMessage, error)
+	Destory()
+
+	PauseSync() error
+	Unpause() error
+
+	APIVersion() (uint32, error)
+	CreateContainer(container string, user *UserGroupInfo, storages []*Storage, c *ocispecs.Spec) error
+	StartContainer(container string) error
+	ExecProcess(container, process string, user *UserGroupInfo, p *ocispecs.Process) error
+	SignalProcess(container, process string, signal syscall.Signal) error
+	WaitProcess(container, process string) int
+
+	// how to handle with these apis when proxy/shim is enabled?
+	WriteStdin(container, process string, data []byte) (int, error)
+	ReadStdout(container, process string, data []byte) (int, error)
+	ReadStderr(container, process string, data []byte) (int, error)
+	CloseStdin(container, process string) error
+	TtyWinResize(container, process string, row, col uint16) error
+
+	StartSandbox(sb *SandboxNetworkConfig, storages []*Storage) error
+	DestroySandbox() error
+	UpdateRoute(r []Route) error
+	UpdateInterface(ifs []Interface) error
+	//OnlineCpuMem() error
+}
+
+// proxies, shim should be implemented via AgentBuilder (and Agent.CreateContainer() Agent.ExecProcess())
+type AgentBuilder interface {
+	Build(AgentConfig) Agent
+}
```

Comment:
The agent package and the sandbox package will be different packages. Therefore, Agent and its Methods have to be exposed so that the sandbox package can use the agent package.